certbot-dns-godaddy/certbot_dns_godaddy/dns_godaddy.py

```diff
@@ -0,0 +1,168 @@
+        :raises certbot.errors.PluginError: if an error occurs communicating with the Godaddy
+                                            API
+        """
+
+        try:
+            domain = self._find_domain(domain_name)
+        except godaddypy.client.BadResponse as e:
+            hint = None
+
+            if "UNABLE_TO_AUTHENTICATE" in str(e):
+                hint = 'Did you provide a valid API token?'
+
+            logger.debug('Error finding domain using the Godaddy API: %s', e)
+            raise errors.PluginError('Error finding domain using the Godaddy API: {0}{1}'
+                                     .format(e, ' ({0})'.format(hint) if hint else ''))
+
+        try:
+            self.client.add_record(domain, {
+                'data': record_content,
+                'name': self._compute_record_name(domain, record_name),
+                'type': 'TXT'
+            })
+            logger.debug('Successfully added TXT record')
+        except godaddypy.client.BadResponse as e:
+            logger.debug('Error adding TXT record using the Godaddy API: %s', e)
+            raise errors.PluginError('Error adding TXT record using the Godaddy API: {0}'
+                                     .format(e))
+
+    def del_txt_record(self, domain_name, record_name, record_content):
+        """
+        Delete a TXT record using the supplied information.
+
+        Note that both the record's name and content are used to ensure that similar records
+        created concurrently (e.g., due to concurrent invocations of this plugin) are not deleted.
+
+        Failures are logged, but not raised.
+
+        :param str domain_name: The domain to use to associate the record with.
+        :param str record_name: The record name (typically beginning with '_acme-challenge.').
+        :param str record_content: The record content (typically the challenge validation).
+        """
+
+        try:
+            domain = self._find_domain(domain_name)
+        except godaddypy.client.BadResponse as e:
+            logger.debug('Error finding domain using the Godaddy API: %s', e)
+            return
+
+        try:
+            domain_records = self.client.get_records(domain, record_type='TXT')
+
+            matching_records = [record for record in domain_records
+                                if record['type'] == 'TXT'
+                                and record['name'] == self._compute_record_name(domain, record_name)
+                                and record['data'] == record_content]
+        except godaddypy.client.BadResponse as e:
+            logger.debug('Error getting DNS records using the Godaddy API: %s', e)
+            return
+
+        for record in matching_records:
+            try:
+                logger.debug('Removing TXT record with name: %s', record['name'])
+                self.client.delete_records(domain, name=record['name'], record_type='TXT')
```

Comment:
It seems like there's a fundamental issue with the GoDaddy API functionality here (i.e., there's no delete operation, just a replace, which is inherently race-y), which we'll probably just have to live with.

It also seems like the GoDaddyPy library could be safer. It does a [full replace](https://github.com/eXamadeus/godaddypy/blob/master/godaddypy/client.py#L277) of all DNS records for the domain during deletion, which could cause data loss if other, potentially unrelated, DNS changes are being made while Certbot is running. Instead, it could do a [limited replace of records that match the type and name](https://developer.godaddy.com/doc#!/_v1_domains/recordReplaceTypeName), which would only impact other ACME challenges.

Better yet would be doing a limited replace _and_ checking the record content in addition to the name and type. Without that, users are likely to encounter issues if multiple instances of Certbot run concurrently.

Would you be interested in working with the library maintainer to see if these changes could be made upstream? If you'd prefer, I'd be happy to reach out to them.